smarts/core/utils/sumo.py

```diff
@@ -41,3 +51,182 @@
+        except traci.exceptions.FatalTraCIError as err:
+            logging.debug("TraCI could not connect in time.")
+            # XXX: the error type is changed to TraCIException to make it consistent with the
+            # process died case of `traci.connect`.
+            raise traci.exceptions.TraCIException(err)
+        except AssertionError:
+            self.close_traci_and_pipes()
+            raise
+
+    @property
+    def sumo_alive(self):
+        """If the SUMO process is alive."""
+        return self._sumo_proc is not None and self._sumo_proc.poll() is None
+
+    def __getattr__(self, name: str) -> Any:
+        if not self.sumo_alive:
+            return None
+
+        attribute = getattr(self._traci_conn, name)
+
+        if inspect.isbuiltin(attribute) or inspect.ismethod(attribute):
+            attribute = functools.partial(
+                _wrap_traci_method, method=attribute, sumo_process=self
+            )
+
+        if isinstance(attribute, traci.domain.Domain):
+            attribute = DomainWrapper(sumo_proc=self, domain=attribute)
+
+        return attribute
+
+    def must_reset(self):
+        """If the version of sumo will have errors if just reloading such that it must be reset."""
+        return self._sumo_version > (1, 12, 0)
+
+    def close_traci_and_pipes(self):
+        """Safely closes all connections. We should expect this method to always work without throwing"""
+
+        def __safe_close(conn):
+            try:
+                conn.close()
+            except (subprocess.SubprocessError, multiprocessing.ProcessError):
+                # Subprocess or process failed
+                pass
+            except traci.exceptions.FatalTraCIError:
+                # TraCI connection is already dead.
+                pass
+            except AttributeError:
+                # Socket was destroyed internally by a fatal error somehow.
+                pass
+
+        if self._traci_conn:
+            __safe_close(self._traci_conn)
+
+        if self._sumo_proc:
+            __safe_close(self._sumo_proc.stdin)
+            __safe_close(self._sumo_proc.stdout)
+            __safe_close(self._sumo_proc.stderr)
+            self._sumo_proc.kill()
+
+        self._sumo_proc = None
+        self._traci_conn = None
+
+    def teardown(self):
+        """Clean up all resources."""
+        self.close_traci_and_pipes()
+
+
+def _wrap_traci_method(*args, method, sumo_process: TraciConn, **kwargs):
+    # Argument order must be `*args` first so keyword arguments are required for `method` and `sumo_process`.
+    try:
+        return method(*args, **kwargs)
+    except traci.exceptions.FatalTraCIError:
+        # Traci cannot continue
+        sumo_process.close_traci_and_pipes()
+        raise
+    except traci.exceptions.TraCIException:
+        # Case where SUMO can continue
+        # TAI: consider closing the process even with a non fatal error
+        pass
```

Comment:
I think this is the one case where I have a question, should we ignore `TraCIException`s?

Reply:
I have decided to re-raise for now.